Add support for the extended attributes of Cookies in the Cookie Store API

RupinMittal opened the pull request:
#### 0fd5d68a2b262b5899042bed3223e53c94f413cd
<pre>
Add support for the extended attributes of Cookies in the Cookie Store API
<a href="https://bugs.webkit.org/show_bug.cgi?id=259340">https://bugs.webkit.org/show_bug.cgi?id=259340</a>

Reviewed by Chris Dumez.

Cookies have more attributes besides just name and value. We want to add
support for gettting/setting these attributes. However, we are not yet
sure if we wish to expose these attributes to JavaScript. So this patch
adds support for the attributes and hides them behind the
CookieStoreAPIExtendedAttributesEnabled feature flag.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/cookie-store/CookieListItem.h:
(WebCore::CookieListItem::CookieListItem):
* Source/WebCore/Modules/cookie-store/CookieListItem.idl:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::get):
(WebCore::CookieStore::getAll):

Canonical link: <a href="https://commits.webkit.org/266171@main">https://commits.webkit.org/266171@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/26fca4d9ae81cd1b2ef4aabdfaf60faddb9642f3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13111 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13425 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13758 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14849 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/12494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13179 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15933 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13452 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15188 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13278 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13963 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11086 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15334 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11251 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11841 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18902 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/11152 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12326 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12009 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15217 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/12407 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12486 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/13161 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11756 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3458 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3205 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16075 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13535 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12331 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3252 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -112,7 +112,7 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
         }
 
         auto& cookie = cookiesVector[0];
```
We don't really need this local variable.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -112,7 +112,7 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
 
         auto& cookie = cookiesVector[0];
-        promise->resolve<IDLDictionary<CookieListItem>>(CookieListItem { WTFMove(cookie.name), WTFMove(cookie.value) });
+        promise->resolve<IDLDictionary<CookieListItem>>(CookieListItem(WTFMove(cookie)));
```
WTFMove(cookiesVector[0])

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -170,9 +170,12 @@ void CookieStore::getAll(CookieStoreGetOptions&& options, Ref<DeferredPromise>&&
             return;
         }
 
-        promise->resolve<IDLSequence<IDLDictionary<CookieListItem>>>(cookies->map([](auto& cookie) {
```
We should try to keep using `map()`, can you try the following?
```
promise->resolve<IDLSequence<IDLDictionary<CookieListItem>>>(WTF::map(WTFMove(cookies), [](auto&& cookie) {
    return CookieListItem { WTFMove(cookie) };
}));
```

RupinMittal replied:
I was trying different ways to use map (I just tried the one above too) but it seems like the lambda that is passed to map must be const and so we aren't allowed to WTFMove the arguments passed in? The above method resulted in some errors too.

cdumez replied:
Pretty sure we can get it to work. We have plenty of existing examples in our code base:
```
LayerAncestorClippingStack::LayerAncestorClippingStack(Vector<CompositedClipData>&& clipDataStack)
    : m_stack(WTF::map(WTFMove(clipDataStack), [](auto&& clipDataEntry) { return ClippingStackEntry { WTFMove(clipDataEntry), 0, nullptr, nullptr }; }))
{
}
``` 

for example.

RupinMittal replied:
Ah alright, I figured out the issue. It should work now.

cdumez reviewed (approved):
LGTM

RupinMittal committed:
Add support for the extended attributes of Cookies in the Cookie Store API
https://bugs.webkit.org/show_bug.cgi?id=259340

Reviewed by Chris Dumez.

Cookies have more attributes besides just name and value. We want to add
support for gettting/setting these attributes. However, we are not yet
sure if we wish to expose these attributes to JavaScript. So this patch
adds support for the attributes and hides them behind the
CookieStoreAPIExtendedAttributesEnabled feature flag.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/cookie-store/CookieListItem.h:
(WebCore::CookieListItem::CookieListItem):
* Source/WebCore/Modules/cookie-store/CookieListItem.idl:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::get):
(WebCore::CookieStore::getAll):

Canonical link: https://commits.webkit.org/266171@main
